### ISSUE TITLE
String literal types don't work when started with `__`

### ISSUE THREAD

**Issue by RReverser** [Microsoft/TypeScript#8614]
**TypeScript Version:** 

nightly (1.9.0-dev.20160515)

**Code**

```ts
var a: '_RESULT' = '_RESULT';
var b: '__RESULT' = '__RESULT';
```

**Expected behavior:**
Both variables are assigned with no errors.

**Actual behavior:** 
`a` is assigned with no errors, but `b` reports type mismatch. This happens whenever I use strings tagged with `__` (double-underscore) as a prefix.

**Comment by DanielRosenwasser** [comment]
Sounds reminiscent of problems we've had with identifier escaping, but it's odd that it pops up here.

**Comment by sandersn** [comment]
The key for string literal types is still escaped with an underscore in this case, so the map looks like this: `{ ___dunder: "__dunder" }`. (The key has three underscores). Naturally, subsequent lookups don't match the key.

**Comment by sandersn** [comment]
1. Intern and escape (all?) strings.
2. Intern identifiers with an escaped key, but store the original text.
3. Unescape identifiers when caching string literal types.

(1) seems like a bad idea. (2) sounds the best to me. (3) is ad-hoc, although it happens to work.

Note that technically any identifier that is used as a key should be escaped because Bad Things will happen if you overwrite `__proto__` in the map object. So if I use (2), I also need to escape string literals when caching them.

**Comment by sandersn** [comment]
(4) Make a separate tagged type for escaped strings called `IdentifierString` or `EscapedIdentifier`. Annotate the entire compiler code with this type. Then decide between (2) and (3). I *really* don't want to do this, and (3) seems good enough for a small issue like this.

**Comment by sandersn** [comment]
Fix is up at #8650.

**Comment by RReverser** [comment]
@sandersn Thanks!
